cpp/include/cudf/lists/detail/scatter.cuh

```diff
@@ -0,0 +1,727 @@
+        ? construct_child_nullmask(
+            list_vector, list_offsets, source_lists, target_lists, num_child_rows, mr, stream)
+        : std::make_pair(rmm::device_buffer{}, 0);
+
+    return cudf::make_lists_column(num_child_rows,
+                                   std::move(child_offsets),
+                                   std::move(child_column),
+                                   child_null_mask.second,            // Null count
+                                   std::move(child_null_mask.first),  // Null mask
+                                   stream,
+                                   mr);
+  }
+};
+
+/**
+ * @brief Checks that the specified columns have matching schemas, all the way down.
+ */
+void assert_same_data_type(column_view const& lhs, column_view const& rhs)
+{
+  CUDF_EXPECTS(lhs.type().id() == rhs.type().id(), "Mismatched Data types.");
+  CUDF_EXPECTS(lhs.num_children() == rhs.num_children(), "Mismatched number of child columns.");
+
+  for (int i{0}; i < lhs.num_children(); ++i) { assert_same_data_type(lhs.child(i), rhs.child(i)); }
+}
+
+}  // namespace
+
+/**
+ * @brief Scatters lists into a copy of the target column
+ * according to a scatter map.
+ *
+ * The scatter is performed according to the scatter iterator such that row
+ * `scatter_map[i]` of the output column is replaced by the source list-row.
+ * All other rows of the output column equal corresponding rows of the target table.
+ *
+ * If the same index appears more than once in the scatter map, the result is
+ * undefined.
+ *
+ * The caller must update the null mask in the output column.
+ *
+ * @tparam SourceIterator must produce list_view objects
+ * @tparam MapIterator must produce index values within the target column.
+ *
+ * @param mr Device memory resource used to allocate the returned column's device memory
+ * @param stream CUDA stream used for device memory operations and kernel launches.
+ * @return New lists column.
+ */
+template <typename MapIterator>
+std::unique_ptr<column> scatter(
+  column_view const& source,
+  MapIterator scatter_map_begin,
+  MapIterator scatter_map_end,
+  column_view const& target,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource(),
+  cudaStream_t stream                 = 0)
+{
+  auto num_rows = target.size();
+
+  if (num_rows == 0) { return cudf::empty_like(target); }
+
+  auto child_column_type = lists_column_view(target).child().type();
+
+  assert_same_data_type(source, target);
+
+  using lists_column_device_view = cudf::detail::lists_column_device_view;
+  using unbound_list_view        = cudf::lists::detail::unbound_list_view;
+
+  auto source_lists_column_view = lists_column_view(source);  // Checks that this is a list column.
+  auto source_device_view       = column_device_view::create(source, stream);
+  auto source_vector            = list_vector_from_column(
+    unbound_list_view::SOURCE, lists_column_device_view(*source_device_view), stream);
```

Comment:
```suggestion
  auto const source_lists_column_view = lists_column_view(source);  // Checks that this is a list column.
  auto const source_device_view       = column_device_view::create(source, stream);
  auto const source_vector            = list_vector_from_column(
    unbound_list_view::SOURCE, lists_column_device_view(*source_device_view), stream);
```

Reply:
Fixed. all the variables except `target_vector` should really be `const`.